{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating Proteomes CAIR\n",
    "\n",
    "* This module is for reading CSV-format files and grouping entries with regard to each organism and calculates the CAIR for each UniProt Organism ID.\n",
    "\n",
    "##### Two output files will be generated:\n",
    "> A) \"All species CAIRs.csv\" containing all Organism IDs, their correspondent CAIRs, and overall residue frequencies.\n",
    "\n",
    "> B) \"Complete proteome CAIRs.csv\" containing non-redundant proteomes and their correspondent CAIRs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from numpy import log2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def entry_to_species(sprot_input_file='Entries sprot.csv', trembl_input_file='Entries trembl.csv',\n",
    "                     outfile='All species residues.csv', chunksize=8000000, merge='True'): # refer to runCAIR\n",
    "    out = pd.DataFrame() # creating an empty datafarme\n",
    "    data = pd.read_csv(trembl_input_file, chunksize=chunksize) # reading data in chunks (to avoid RAM insufficiency)\n",
    "\n",
    "    for chunk in data:\n",
    "        chunk = chunk.drop(columns=['CAIR']) # removing CAIRs for each protein (not needed anymore)\n",
    "        group = chunk.groupby('Organism_ID').sum() # grouping by organisms for each chunk\n",
    "        out = out.append(group) # filling out the dataframe\n",
    "\n",
    "    if merge == 'True': # whether to merge the Swiss-Prot file to the prepared TrEMBL file or not\n",
    "        data2 = pd.read_csv(sprot_input_file).drop(columns=['CAIR']) # reading the Swiss-Prot file\n",
    "        data2 = data2.groupby('Organism_ID').sum() # grouping by organisms\n",
    "        out = pd.concat([out, data2]) # adding the Swiss-Prot to the TrEMBL\n",
    "\n",
    "    out = out.groupby('Organism_ID').sum() # grouping by organisms\n",
    "    out.to_csv(outfile) # writing the CSV file\n",
    "    \n",
    "\n",
    "def species_cair(input_file='All species residues.csv', outfile='Complete proteome CAIRs.csv', proteomes_file=\"proteomes-redundant_no.tab\"): # refer to runCAIR\n",
    "    data = pd.DataFrame(pd.read_csv(input_file, dtype=float)) # reading the input\n",
    "    length = data.Len # defining lengths for frequency calculations\n",
    "    cair = pd.DataFrame((log2((data[\"A\"] / length) ** (-data[\"A\"] / length)) + log2(\n",
    "        (data[\"C\"] / length) ** (-data[\"C\"] / length)) + log2(\n",
    "        (data[\"D\"] / length) ** (-data[\"D\"] / length)) + log2(\n",
    "        (data[\"E\"] / length) ** (-data[\"E\"] / length)) + log2((data[\"F\"] / length) ** (-data[\"F\"] / length)) + log2(\n",
    "        (data[\"G\"] / length) ** (-data[\"G\"] / length)) + log2(\n",
    "        (data[\"H\"] / length) ** (-data[\"H\"] / length)) + log2((data[\"I\"] / length) ** (-data[\"I\"] / length)) + log2(\n",
    "        (data[\"K\"] / length) ** (-data[\"K\"] / length)) + log2(\n",
    "        (data[\"L\"] / length) ** (-data[\"L\"] / length)) + log2((data[\"M\"] / length) ** (-data[\"M\"] / length)) + log2(\n",
    "        (data[\"N\"] / length) ** (-data[\"N\"] / length)) + log2(\n",
    "        (data[\"O\"] / length) ** (-data[\"O\"] / length)) + log2((data[\"P\"] / length) ** (-data[\"P\"] / length)) + log2(\n",
    "        (data[\"Q\"] / length) ** (-data[\"Q\"] / length)) + log2(\n",
    "        (data[\"R\"] / length) ** (-data[\"R\"] / length)) + log2((data[\"S\"] / length) ** (-data[\"S\"] / length)) + log2(\n",
    "        (data[\"T\"] / length) ** (-data[\"T\"] / length)) + log2(\n",
    "        (data[\"U\"] / length) ** (-data[\"U\"] / length)) + log2((data[\"V\"] / length) ** (-data[\"V\"] / length)) + log2(\n",
    "        (data[\"W\"] / length) ** (-data[\"W\"] / length)) + log2((data[\"Y\"] / length) ** (-data[\"Y\"] / length))) / log2(22)) # calculating CAIRs\n",
    "    cair.columns = [\"CAIR\"] # assigning the column name\n",
    "\n",
    "    Organism_ID = pd.DataFrame(data.Organism_ID)\n",
    "    Organism_CAIR = Organism_ID.join(cair)\n",
    "    Proteomes = pd.read_csv(proteomes_file, sep=\"\\t\") # reading the proteomes list file\n",
    "    Sp_CAIR = Organism_CAIR.merge(Proteomes, left_on=\"Organism_ID\", right_on=\"Organism ID\").drop_duplicates('Organism_ID',\n",
    "                                                                                                          keep='last') # merging taxonomy data with species CAIRs\n",
    "    Sp_CAIR['First_hierarchy'] = Sp_CAIR['Taxonomic lineage'].str.split(', ', expand=True)[0] # extracting first hierarchy of organisms (superkingdoms)\n",
    "    Sp_CAIR['Taxonomic lineage'] = Sp_CAIR['Taxonomic lineage'].str.split(', ', expand=True)[1] # extracting second hierarchy of organisms (phyla in most cases)\n",
    "    Sp_CAIR = Sp_CAIR[Sp_CAIR.First_hierarchy != 'Viruses'].drop(columns=['First_hierarchy', 'Proteome ID', 'Organism ID']).rename(\n",
    "        columns={'Taxonomic lineage': 'Second_hierarchy'}).reset_index(drop=True) # filtering viruses out and dropping unnecessary columns and renaming\n",
    "\n",
    "    code = [] # creating an empty list for analysis codes\n",
    "    for row in Sp_CAIR['Second_hierarchy']: # assining codes according to the tree of life\n",
    "        if row == 'Proteobacteria':\n",
    "            code.append('1')\n",
    "        elif row == 'Candidatus Hydrogenedentes':\n",
    "            code.append('2')\n",
    "        elif row == 'Candidatus Abyssubacteria':\n",
    "            code.append('3')\n",
    "        elif row == 'Spirochaetes':\n",
    "            code.append('4')\n",
    "        elif row == 'Deferribacteres':\n",
    "            code.append('5')\n",
    "        elif row == 'Chrysiogenetes':\n",
    "            code.append('6')\n",
    "        elif row == 'Acidobacteria':\n",
    "            code.append('7')\n",
    "        elif row == 'Thermodesulfobacteria':\n",
    "            code.append('8')\n",
    "        elif row == 'Nitrospirae':\n",
    "            code.append('9')\n",
    "        elif row == 'Nitrospinae/Tectomicrobia group':\n",
    "            code.append('10')\n",
    "        elif row == 'Elusimicrobia':\n",
    "            code.append('11')\n",
    "        elif row == 'Candidatus Omnitrophica':\n",
    "            code.append('12')\n",
    "        elif row == 'Planctomycetes':\n",
    "            code.append('13')\n",
    "        elif row == 'Chlamydiae':\n",
    "            code.append('14')\n",
    "        elif row == 'Lentisphaerae':\n",
    "            code.append('15')\n",
    "        elif row == 'Candidatus Aureabacteria':\n",
    "            code.append('16')\n",
    "        elif row == 'Kiritimatiellaeota':\n",
    "            code.append('17')\n",
    "        elif row == 'Verrucomicrobia':\n",
    "            code.append('18')\n",
    "        elif row == 'Candidatus Aegiribacteria':\n",
    "            code.append('19')\n",
    "        elif row == 'Candidatus Latescibacteria':\n",
    "            code.append('20')\n",
    "        elif row == 'Gemmatimonadetes':\n",
    "            code.append('21')\n",
    "        elif row == 'Candidatus Fermentibacteria':\n",
    "            code.append('22')\n",
    "        elif row == 'Candidatus Marinimicrobia':\n",
    "            code.append('23')\n",
    "        elif row == 'candidate division LCP-89':\n",
    "            code.append('24')\n",
    "        elif row == 'Calditrichaeota':\n",
    "            code.append('25')\n",
    "        elif row == 'Rhodothermaeota':\n",
    "            code.append('26')\n",
    "        elif row == 'Balneolaeota':\n",
    "            code.append('27')\n",
    "        elif row == 'Ignavibacteriae':\n",
    "            code.append('28')\n",
    "        elif row == 'Candidatus Kryptonia':\n",
    "            code.append('29')\n",
    "        elif row == 'Chlorobi':\n",
    "            code.append('30')\n",
    "        elif row == 'Bacteroidetes':\n",
    "            code.append('31')\n",
    "        elif row == 'Candidatus Kapabacteria':\n",
    "            code.append('32')\n",
    "        elif row == 'Candidatus Cloacimonetes':\n",
    "            code.append('33')\n",
    "        elif row == 'Fibrobacteres':\n",
    "            code.append('34')\n",
    "        elif row == 'Synergistetes':\n",
    "            code.append('35')\n",
    "        elif row == 'Fusobacteria':\n",
    "            code.append('36')\n",
    "        elif row == 'Deinococcus-Thermus':\n",
    "            code.append('37')\n",
    "        elif row == 'Coprothermobacterota':\n",
    "            code.append('38')\n",
    "        elif row == 'Thermotogae':\n",
    "            code.append('39')\n",
    "        elif row == 'Aquificae':\n",
    "            code.append('40')\n",
    "        elif row == 'Caldiserica/Cryosericota group':\n",
    "            code.append('41')\n",
    "        elif row == 'Dictyoglomi':\n",
    "            code.append('42')\n",
    "        elif row == 'Firmicutes':\n",
    "            code.append('43')\n",
    "        elif row == 'Tenericutes':\n",
    "            code.append('44')\n",
    "        elif row == 'Candidatus Eremiobacteraeota':\n",
    "            code.append('45')\n",
    "        elif row == 'Abditibacteriota':\n",
    "            code.append('46')\n",
    "        elif row == 'Armatimonadetes':\n",
    "            code.append('47')\n",
    "        elif row == 'Thermobaculum':\n",
    "            code.append('48')\n",
    "        elif row == 'Chloroflexi':\n",
    "            code.append('49')\n",
    "        elif row == 'Candidatus Dormibacteraeota':\n",
    "            code.append('50')\n",
    "        elif row == 'Actinobacteria':\n",
    "            code.append('51')\n",
    "        elif row == 'Cyanobacteria':\n",
    "            code.append('52')\n",
    "        elif row == 'Candidatus Melainabacteria':\n",
    "            code.append('53')\n",
    "        elif row == 'Candidatus Margulisbacteria':\n",
    "            code.append('54')\n",
    "        elif row == 'Candidatus Saganbacteria':\n",
    "            code.append('55')\n",
    "        elif row == 'Candidatus Saccharibacteria' or\\\n",
    "                row == 'candidate division SR1' or\\\n",
    "                row == 'Candidatus Dependentiae' or\\\n",
    "                row == 'Candidatus Gracilibacteria' or\\\n",
    "                row == 'Candidatus Atribacteria' or\\\n",
    "                row == 'Candidatus Parcubacteria' or\\\n",
    "                row == 'Candidatus Bipolaricaulota' or\\\n",
    "                row == 'Candidatus Poribacteria' or\\\n",
    "                row == 'unclassified Parcubacteria group' or\\\n",
    "                row == 'Candidatus Aminicenantes' or\\\n",
    "                row == 'Candidatus Coatesbacteria' or\\\n",
    "                row == 'Candidatus Eisenbacteria' or\\\n",
    "                row == 'Candidatus Aerophobetes' or\\\n",
    "                row == 'Candidatus Riflebacteria' or\\\n",
    "                row == 'Candidatus Wolfebacteria' or\\\n",
    "                row == 'Candidatus Nomurabacteria' or\\\n",
    "                row == 'Candidatus Roizmanbacteria' or\\\n",
    "                row == 'Candidatus Uhrbacteria' or\\\n",
    "                row == 'Candidatus Yanofskybacteria' or\\\n",
    "                row == 'Candidatus Levybacteria' or\\\n",
    "                row == 'Candidatus Colwellbacteria' or\\\n",
    "                row == 'candidate division WWE3' or\\\n",
    "                row == 'Candidatus Sungbacteria' or\\\n",
    "                row == 'Candidatus Woykebacteria' or\\\n",
    "                row == 'Candidatus Komeilibacteria' or\\\n",
    "                row == 'Candidatus Falkowbacteria' or\\\n",
    "                row == 'Candidatus Moranbacteria' or\\\n",
    "                row == 'Candidatus Curtissbacteria' or\\\n",
    "                row == 'Candidatus Giovannonibacteria' or\\\n",
    "                row == 'Candidatus Dojkabacteria' or\\\n",
    "                row == 'Candidatus Harrisonbacteria' or\\\n",
    "                row == 'Candidatus Magasanikbacteria' or\\\n",
    "                row == 'Candidatus Beckwithbacteria' or\\\n",
    "                row == 'Candidatus Fraserbacteria' or\\\n",
    "                row == 'Candidatus Kaiserbacteria' or\\\n",
    "                row == 'Candidatus Niyogibacteria' or\\\n",
    "                row == 'Candidatus Yonathbacteria' or\\\n",
    "                row == 'Candidatus Azambacteria' or\\\n",
    "                row == 'Candidatus Portnoybacteria' or\\\n",
    "                row == 'Candidatus Berkelbacteria' or\\\n",
    "                row == 'Candidatus Doudnabacteria' or\\\n",
    "                row == 'Candidatus Zambryskibacteria' or\\\n",
    "                row == 'Candidatus Staskawiczbacteria' or\\\n",
    "                row == 'Candidatus Woesebacteria' or\\\n",
    "                row == 'Candidatus Lloydbacteria' or\\\n",
    "                row == 'Candidatus Nealsonbacteria' or\\\n",
    "                row == 'Candidatus Microgenomates' or\\\n",
    "                row == 'Candidatus Taylorbacteria' or\\\n",
    "                row == 'Candidatus Vogelbacteria' or\\\n",
    "                row == 'Candidatus Buchananbacteria' or\\\n",
    "                row == 'Candidatus Gottesmanbacteria' or\\\n",
    "                row == 'Candidatus Jorgensenbacteria' or\\\n",
    "                row == 'Candidatus Rokubacteria' or\\\n",
    "                row == 'Candidatus Peregrinibacteria' or\\\n",
    "                row == 'Candidatus Dadabacteria' or\\\n",
    "                row == 'Candidatus Kerfeldbacteria' or\\\n",
    "                row == 'Candidatus Desantisbacteria' or\\\n",
    "                row == 'Candidatus Ryanbacteria' or\\\n",
    "                row == 'Candidatus Pacebacteria' or\\\n",
    "                row == 'Candidatus Daviesbacteria' or\\\n",
    "                row == 'Candidatus Amesbacteria' or\\\n",
    "                row == 'Candidatus Tagabacteria' or\\\n",
    "                row == 'Candidatus Shapirobacteria' or\\\n",
    "                row == 'candidate division CPR1' or\\\n",
    "                row == 'Candidatus Adlerbacteria' or\\\n",
    "                row == 'Candidatus Spechtbacteria' or\\\n",
    "                row == 'candidate division Kazan-3B-28' or\\\n",
    "                row == 'Candidatus Terrybacteria' or\\\n",
    "                row == 'Candidatus Wildermuthbacteria' or\\\n",
    "                row == 'candidate division NC10' or\\\n",
    "                row == 'Candidatus Campbellbacteria' or\\\n",
    "                row == 'Candidatus Collierbacteria' or\\\n",
    "                row == 'Candidatus Wirthbacteria' or\\\n",
    "                row == 'Candidatus Brennerbacteria' or\\\n",
    "                row == 'Candidatus Kuenenbacteria' or\\\n",
    "                row == 'Candidatus Veblenbacteria' or\\\n",
    "                row == 'candidate division KSB1' or\\\n",
    "                row == 'Candidatus Glassbacteria' or\\\n",
    "                row == 'Candidatus Firestonebacteria' or\\\n",
    "                row == 'Candidatus Delongbacteria' or\\\n",
    "                row == 'Candidatus Lindowbacteria' or\\\n",
    "                row == 'candidate division TA06' or\\\n",
    "                row == 'Candidatus Liptonbacteria' or\\\n",
    "                row == 'Candidatus Jacksonbacteria' or\\\n",
    "                row == 'Candidatus Blackburnbacteria' or\\\n",
    "                row == 'Candidatus Abawacabacteria' or\\\n",
    "                row == 'Candidatus Wallbacteria' or\\\n",
    "                row == 'Candidatus Schekmanbacteria' or\\\n",
    "                row == 'Candidatus Hydrothermae' or\\\n",
    "                row == 'candidate division WOR-3' or\\\n",
    "                row == 'Candidatus Sumerlaeota' or\\\n",
    "                row == 'candidate division KSB3' or\\\n",
    "                row == 'Candidatus Andersenbacteria' or\\\n",
    "                row == 'candidate division WS5' or\\\n",
    "                row == 'Candidatus Edwardsbacteria' or\\\n",
    "                row == 'Candidatus Chisholmbacteria' or\\\n",
    "                row == 'Candidatus Fischerbacteria' or\\\n",
    "                row == 'candidate division KD3-62' or\\\n",
    "                row == 'candidate division CPR3' or\\\n",
    "                row == 'Candidatus Handelsmanbacteria' or\\\n",
    "                row == 'candidate division CPR2' or\\\n",
    "                row == 'Candidatus Cerribacteria' or\\\n",
    "                row == 'Candidatus Raymondbacteria' or\\\n",
    "                row == 'Candidatus Goldbacteria':\n",
    "            code.append('56') # Assigning one code for all Candidate Phyla Radiation(CPR)\n",
    "        elif row == 'Candidatus Hydrothermarchaeota':\n",
    "            code.append('57')\n",
    "        elif row == 'Candidatus Altiarchaeota':\n",
    "            code.append('58')\n",
    "        elif row == 'Candidatus Micrarchaeota':\n",
    "            code.append('59')\n",
    "        elif row == 'Candidatus Diapherotrites':\n",
    "            code.append('60')\n",
    "        elif row == 'Candidatus Aenigmarchaeota':\n",
    "            code.append('61')\n",
    "        elif row == 'Candidatus Huberarchaea':\n",
    "            code.append('62')\n",
    "        elif row == 'Nanoarchaeota':\n",
    "            code.append('63')\n",
    "        elif row == 'Candidatus Parvarchaeota':\n",
    "            code.append('64')\n",
    "        elif row == 'Candidatus Pacearchaeota':\n",
    "            code.append('65')\n",
    "        elif row == 'Candidatus Woesearchaeota':\n",
    "            code.append('66')\n",
    "        elif row == 'Euryarchaeota':\n",
    "            code.append('67')\n",
    "        elif row == 'Candidatus Bathyarchaeota':\n",
    "            code.append('68')\n",
    "        elif row == 'Thaumarchaeota':\n",
    "            code.append('69')\n",
    "        elif row == 'Candidatus Geothermarchaeota':\n",
    "            code.append('70')\n",
    "        elif row == 'Candidatus Korarchaeota':\n",
    "            code.append('71')\n",
    "        elif row == 'Candidatus Nezhaarchaeota':\n",
    "            code.append('72')\n",
    "        elif row == 'Candidatus Verstraetearchaeota':\n",
    "            code.append('73')\n",
    "        elif row == 'Candidatus Marsarchaeota':\n",
    "            code.append('74')\n",
    "        elif row == 'Crenarchaeota':\n",
    "            code.append('75')\n",
    "        elif row == 'Asgard group':\n",
    "            code.append('76')\n",
    "        elif row == 'Euglenozoa':\n",
    "            code.append('77')\n",
    "        elif row == 'Heterolobosea':\n",
    "            code.append('78')\n",
    "        elif row == 'Metamonada':\n",
    "            code.append('79')\n",
    "        elif row == 'Apusozoa':\n",
    "            code.append('80')\n",
    "        elif row == 'Rotosphaerida':\n",
    "            code.append('81')\n",
    "        elif row == 'Fungi':\n",
    "            code.append('82')\n",
    "        elif row == 'Ichthyosporea':\n",
    "            code.append('83')\n",
    "        elif row == 'Filasterea':\n",
    "            code.append('84')\n",
    "        elif row == 'Choanoflagellata':\n",
    "            code.append('85')\n",
    "        elif row == 'Metazoa':\n",
    "            code.append('86')\n",
    "        elif row == 'Amoebozoa':\n",
    "            code.append('87')\n",
    "        elif row == 'Haptista':\n",
    "            code.append('88')\n",
    "        elif row == 'Cryptophyceae':\n",
    "            code.append('89')\n",
    "        elif row == 'Sar':\n",
    "            code.append('90')\n",
    "        elif row == 'Viridiplantae':\n",
    "            code.append('91')\n",
    "        elif row == 'Rhodophyta':\n",
    "            code.append('92')\n",
    "\n",
    "        else:\n",
    "            code.append(\"0\") # the following six phyla were not matched, thus excluded:\n",
    "                             # Haloplasmatales, environmental samples, Candidatus Vecturithrix, Vampirococcus, Natronospirillum, unclassified DPANN group\n",
    "\n",
    "    Sp_CAIR['code'] = code # inserting the code column\n",
    "    Sp_CAIR.to_csv(outfile, index=False) # writing the CSV output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
